### DISCUSSION TITLE
Using UserEventRequest in Alexa.NET

### DISCUSSION THREAD

**Discussion by abpatel** [stoiveyp/Alexa.NET.APL#94] [Q&A] [answered]
Is there a code snippet that illustrates how to use  UserEventRequest?
I noticed this in the docs:
`new UserEventRequestHandler ().AddToRequestConverter();`

But where exactly do I add this? Do I just add that single line in the constructor of the lambda?
Basically I'm looking to handle a user press on an AlexaGridList item to receive the item/ordinal of the data item and in response send an updated data source back to the device. 
1) Do I send back a full document with the updated data source? 
OR
2) is it possible to send back just the updated data item back to the device. If so how do I do this in Alexa.Net

A complete code snippet of a lambda that handles  a UserEvent would be very helpful.

Thank you.

**Comment by stoiveyp** [comment, marked as answer]
Hi @abpatel 

You add the `AddToRequestConverter` code once at startup (or in your function constructor if you're using a serverless pattern like AWS Lambda). This is an Alexa.NET pattern we use with the extensions, so when you look at the `skillRequest.Request` property on your input, it can correctly deserialize a `UserEventRequest` type when you receive one from your APL document (it doesn't break if you run it more than once, but it doesn't need to be run more than once at startup)

As for the response you send back - it depends on your skill setup. You can send almost any kind of response back from a UserEvent request: speech, a question, a new screen etc.

You've said that in your case you want to change the data source depending on the selection, so having google'd the Alexa documentation - you'd need to ensure you're using a `DynamicIndexList` data source and then respond with an `UpdateIndexListDataDirective` directive added to your response.

The documentation that came back when I googled "APL update data source" is below, and I try and make sure the names in the library are close enough you can always find them with a quick search - that way the Alexa docs are still the best way to figure out how to do something (some scenarios are just so specific I wouldn't be able to keep all the examples up to date)

https://developer.amazon.com/en-US/docs/alexa/alexa-presentation-language/apl-data-source.html#use-directives-and-requests-to-manage-the-list

Hope this helps

**Reply by abpatel** [reply]
Thank you!  `DynamicIndexList` did the trick!
Worked beautifully for my use case since I needed to send updates to my pre bound data source.
Thank you for the guidance.
